opm/parser/eclipse/EclipseState/Schedule/Well.cpp

```diff
@@ -303,11 +526,109 @@ namespace Opm {
+                        const double distance = (*currentSegmentSet)[i_segment]->length();
+                        const double distance_difference = std::abs(center_distance - distance);
+                        if (distance_difference < min_distance_difference) {
+                            min_distance_difference = distance_difference;
+                            segment_number = (*currentSegmentSet)[i_segment]->segmentNumber();
+                        }
+                    }
+                }
+
+                if (segment_number != 0) {
+                    compsegs[i_comp]->m_segment_number = segment_number;
+                    if (compsegs[i_comp]->m_center_depth == 0.) {
+                        // using the depth of the segment node as the depth of the completion
+                        // TODO: now only one completion for one segment is hanlded,
+                        // TODO: later we will try to handle more than one completion for each segment,
+                        // TODO: which will be a linear interpolation based on the segment node depth
+                        // TODO: in the same branch.
+                        int segment_location = currentSegmentSet->numberToLocation(segment_number);
+                        compsegs[i_comp]->m_center_depth = (*currentSegmentSet)[segment_location]->depth();
+                    }
+                } else {
+                   throw std::runtime_error(" the pefomration failed in finding a segment \n");
+                }
+            }
+        }
+
+
+        CompletionSetConstPtr currentCompletionSet = m_completions->get(time_step);
+        // it is necessary to update the segment related information for some completions.
+        CompletionSetPtr newCompletionSet = CompletionSetPtr(currentCompletionSet->shallowCopy());
+
+
+        for (size_t i_comp = 0; i_comp < compsegs.size(); ++i_comp) {
+            const int i = compsegs[i_comp]->m_i;
+            const int j = compsegs[i_comp]->m_j;
+            const int k = compsegs[i_comp]->m_k;
+
+            size_t ic;
+            for (ic = 0; ic < newCompletionSet->size(); ++ic) {
+                if (newCompletionSet->get(ic)->sameCoordinate(i, j, k)) {
+                    break; // the completion is found
+                }
+            }
+
+            if (ic == newCompletionSet->size()) {
+                throw std::runtime_error(" the completion specified in COMPSEGS is not found in the completionSet \n");
+            }
+
+            CompletionPtr new_completion = std::make_shared<Completion>(newCompletionSet->get(ic));
+            new_completion->setSegmentNumber(compsegs[i_comp]->m_segment_number);
+            new_completion->setCenterDepth(compsegs[i_comp]->m_center_depth);
+            newCompletionSet->add(new_completion);
+        }
+
+        for (size_t ic = 0; ic < newCompletionSet->size(); ++ic) {
+            if (newCompletionSet->get(ic)->getSegmentNumber() == -1) {
```

Comment:
Can we instead have a new method `bool Completion::attachedToSegment()` - I strongly dislike spreading arbitrary invalid values like '-1' to the rest of the code.